### DISCUSSION TITLE
Edge case where derived atoms that depend on stable derived atoms are re-computed

### DISCUSSION THREAD

**Discussion by Pinpickle** [pmndrs/jotai#2738] [Bug report]
### Bug Description

This is a weird one... I've managed to find a situation where atoms that depend on stable derived atoms (that is, they return a stable value) end up re-evaluating.

Rather than try to explain what's going on, here's a failing test case:

```ts
it('should not re-evaluate stable derived atom values in situations where dependencies are re-ordered', () => {
  const callCounter = vitest.fn()
  const rootAtom = atom(false);
  const stableDep = atom(get => {
    get(rootAtom);
    return 1;
  });
  const stableDepDep = atom(get => {
    get(stableDep);
    callCounter();
    return 2;
  });

  const newAtom = atom(get => {
    if (get(rootAtom)) {
      return get(stableDepDep);
    };

    return get(stableDep);
  });
  
  const store = createStore();
  store.sub(stableDepDep, () => {});
  store.sub(newAtom, () => {});
  expect(store.get(stableDepDep)).toBe(2);
  expect(callCounter).toHaveBeenCalledTimes(1);

  store.set(rootAtom, true);
  expect(store.get(newAtom)).toBe(2);
  expect(callCounter).toHaveBeenCalledTimes(1);
});
```

From what I can tell, it's due to the combination of the dependency order changing and the `force` being passed by `recomputeDependents`: https://github.com/pmndrs/jotai/blob/b4158262d376228bc488991c7616ad2b84d810bf/src/vanilla/store.ts#L548

### Reproduction Link

https://stackblitz.com/github/Pinpickle/jotai/tree/stable-dependency-bug

**Comment by dai-shi** [comment]
Thanks for reporting.
Do you have any idea for the solution?

**Reply by Pinpickle** [reply]
No concrete ideas! I'm not much of an expert on reactive programming...

I think `recomputeDependents` would need to notice if it is recomputing an atom that was previously a dependency but hasn't yet been re-evaluated in topological order (so it's still in the `isMarked` set by the time we arrive at a get). At that point, it'd need to change the order it recomputes dependencies in... which is the point I'm getting stuck at.

**Reply by dai-shi** [reply]
Thanks. I understand the problem too.

PR: #2742
